### PR TITLE
feat: implement fuse-t feature

### PR DESCRIPTION
**Pull request by killagu**
Use fuse-t to replace MacFUSE as the implementation of FUSE. fuse-t is based on NFS and does not require the use of kernel extensions. Installation can be done with lower privileges and it does not impact system stability.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -51,6 +51,7 @@ async-io = ["async-trait", "tokio-uring", "tokio/fs", "tokio/net", "tokio/sync",
 fusedev = ["vmm-sys-util", "caps", "core-foundation-sys"]
 virtiofs = ["virtio-queue", "caps", "vmm-sys-util"]
 vhost-user-fs = ["virtiofs", "vhost", "caps"]
+fuse-t=[]
```
Comment (bergwolf) [inline Cargo.toml:54]: should be a feature only available on macos?

File: src/transport/fusedev/mod.rs [hunk 2]
```diff
@@ -28,10 +28,19 @@ mod linux_session;
 pub use linux_session::*;
 
 #[cfg(target_os = "macos")]
+#[cfg(not(feature = "fuse-t"))]
```
Comment (bergwolf) [inline src/transport/fusedev/mod.rs:31]: `#[cfg(all(target_os = "macos", not(feature = "fuse-t")))]`?

File: tests/macfuse_smoke.rs [hunk 3]
```diff
@@ -53,7 +53,7 @@ mod macfuse_tests {
     }
 
     #[test]
-    #[ignore] // it depends on privileged mode to pass through /dev/fuse
+    #[cfg(feature = "fuse-t")] // it depends on privileged mode to pass through /dev/fuse
```
Comment (bergwolf) [inline tests/macfuse_smoke.rs:56]: we use `--ignore` to test with privilege mode.

**Review by bergwolf — COMMENTED** [review]
Now that we have a third fuse session implementation, IMHO we should define a trait for all of them. But we can do it later on.

File: tests/macfuse_smoke.rs (hunk 3 above)
Comment (killagu) [reply tests/macfuse_smoke.rs:56]: This case only run macos, has no `--ignore` run.

Comment (killagu) [reply tests/macfuse_smoke.rs:56]: macfuse can not get privilege, becase it must click in `System Setting`.

File: src/transport/fusedev/fuse_t_session.rs [hunk 4]
```diff
@@ -0,0 +1,638 @@
...
+                .map_err(|e| SessionFailure(format!("parent close mon fd0 failed {:?}", e)))?;
+            unsafe { Ok((File::from_raw_fd(fd1), File::from_raw_fd(mon_fd1))) }
+        }
+        ForkResult::Child => {
+            close(fd1).map_err(|e| SessionFailure(format!("child close fd1 failed {:?}", e)))?;
+            close(mon_fd1)
+                .map_err(|e| SessionFailure(format!("child close mon fd1 failed {:?}", e)))?;
+
+            let mut daemon_path: Vec<u8> = Vec::with_capacity(PROC_PIDPATHINFO_MAXSIZE as usize);
+            unsafe {
```
Comment (bergwolf) [inline src/transport/fusedev/fuse_t_session.rs:387]: please limit the scope of `unsafe` to just `proc_pidpath`. IIUC, it is the only one that requires `unsafe`, right?

File: src/transport/fusedev/fuse_t_session.rs [hunk 5]
```diff
@@ -0,0 +1,638 @@
...
+
+use super::{Error::IoError, Error::SessionFailure, FuseBuf, FuseDevWriter, Reader, Result};
+use crate::transport::pagesize;
+
+// These follows definition from libfuse.
+const FUSE_KERN_BUF_SIZE: usize = 256;
+const FUSE_HEADER_SIZE: usize = 0x1000;
+const FS_SND_SIZE: usize = 4 * 1024 * 1024;
+
+const FUSE_NFSSRV_PATH: &str = "/usr/local/bin/go-nfsv4";
```
[the remark this answers is not in the document]
Comment (killagu) [reply src/transport/fusedev/fuse_t_session.rs:36]: Install by this script
```shell
wget https://github.com/macos-fuse-t/fuse-t/releases/download/1.0.24/fuse-t-macos-installer-1.0.24.pkg
sudo installer -pkg fuse-t-macos-installer-1.0.24.pkg -target /
```

File: src/transport/fusedev/fuse_t_session.rs (hunk 4 above)
Comment (killagu) [reply src/transport/fusedev/fuse_t_session.rs:388]: And `daemon_path.set_len` is unsafe too.

**Comment by bergwolf** [comment]
There are a few warnings:
```
warning: unused import: `vmm_sys_util::tempdir::TempDir`
  --> tests/macfuse_smoke.rs:17:9
   |
17 |     use vmm_sys_util::tempdir::TempDir;
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(unused_imports)]` on by default

warning: unused import: `crate::example::macfuse`
  --> tests/macfuse_smoke.rs:19:9
   |
19 |     use crate::example::macfuse;
   |         ^^^^^^^^^^^^^^^^^^^^^^^

warning: function `validate_hello_file` is never used
  --> tests/macfuse_smoke.rs:21:8
   |
21 |     fn validate_hello_file(dest: &str) -> bool {
   |        ^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(dead_code)]` on by default

warning: function `exec` is never used
  --> tests/macfuse_smoke.rs:38:8
   |
38 |     fn exec(cmd: &str) -> Result<String> {
   |        ^^^^

warning: `fuse-backend-rs` (test "macfuse_smoke") generated 4 warnings (run `cargo fix --test "macfuse_smoke"` to apply 2 suggestions)
```
Not sure why CI didn't catch it.

Otherwise lgtm!

**Comment by killagu** [comment]
`tests/macfuse_smoke.rs` this file only run `fuse-t` enabled, so has warning in `fusedev` test.

Add `feature = fuse-t`  works.

**Review by bergwolf — APPROVED** [review]
lgtm, thanks!

**Pull request merged**